datalad/cmd.py

```diff
@@ -6,66 +6,86 @@
 #   copyright and license terms.
 #
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command calls, allowing for dry runs and output handling
+
+"""
 
-import os
-import commands
 
+import subprocess
+import sys
 import logging
-lgr = logging.getLogger('datalad.files')
+import os
+
+lgr = logging.getLogger('datalad.cmd')
 
-def dry(s, dry_):
-    """Small helper for dry runs
-    """
-    if dry_:
-        return "DRY " + s
-    return s
 
 class Runner(object):
-    """Helper to run commands and functions while allowing for dry_run and later reporting
+    """Provides a wrapper for system calls.
+
+    An object of this class provides a method to make system calls.
+    Outputs (stdout and stderr) are streamed to stdout during execution as if you were calling it from command line.
+    Additionally allows for dry runs. This is achieved by initializing the `Runner` with `dry=True`.
+    The Runner will then collect all calls as strings in `commands`.
     """
 
     __slots__ = ['commands', 'dry']
 
     def __init__(self, dry=False):
+        self.dry = dry
         self.commands = []
-        self.dry = dry                    # TODO: make use of it
 
-    # TODO -- remove dry_run specification here -- use constructor parameter
-    #def __call__(self, cmd, dry_run=False):
-    def getstatusoutput(self, cmd, dry_run=None):
-        """A wrapper around commands.getstatusoutput
+    def run(self, cmd):
+        """Runs the command `cmd` using shell.
+
+        `cmd` is called and uses stdout whereas stderr is captured and logged.
+        In case of dry-mode `cmd` is just added to `commands`.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        Returns
+        -------
+        Status code as returned by the called command or `None` in case of dry-mode.
         """
-        self.log("Running: %s" % (cmd,))
+
+        self.log("Running: %s" % cmd)
+
         if not self.dry:
-            # status, output = commands.getstatusoutput(cmd)
-            # doing manually for improved debugging
-            pipe = os.popen('{ ' + cmd + '; } 2>&1', 'r')
-            output = ''
-            for line in iter(pipe.readline, ''):
-                self.log("| " + line.rstrip('\n'), level=5)
-                output += line
-            if output[-1:] == '\n': output = output[:-1]
-            status = pipe.close() or 0
+
+            proc = subprocess.Popen(cmd, stdout=sys.stdout, stderr=subprocess.PIPE, shell=True)
```

Comment:
A bit confused... output should now come from this popen and get into the logger as before, and formatted. Now only with separated stdout and stderr

On March 8, 2015 5:23:04 PM EDT, bpoldrack <notifications@github.com> wrote:
>>          if not self.dry:
>> -            # status, output = commands.getstatusoutput(cmd)
>> -            # doing manually for improved debugging
>> -            pipe = os.popen('{ ' + cmd + '; } 2>&1', 'r')
>> -            output = ''
>> -            for line in iter(pipe.readline, ''):
>> -                self.log("| " + line.rstrip('\n'), level=5)
>> -                output += line
>> -            if output[-1:] == '\n': output = output[:-1]
>> -            status = pipe.close() or 0
>> +
>> +            proc = subprocess.Popen(cmd, stdout=sys.stdout,
>stderr=subprocess.PIPE, shell=True)
>
>Ok, agree to the compromise. But by "just logged" do you mean the use
>of the logger or your former way to get the output? And do you want
>that formatting for stderr-logs too?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/datalad/datalad/pull/53/files#r26009159